File: bigbase/src/models/Dispatch.js
export default class Dispatch {
    constructor(dispatchState) {
        let date = new Date().toLocaleDateString('en-GB').split("/").map(x => {
            let n = x.toString();
            n = n.length < 2 ? `0${n}` : n;
            return n
        })
        date = [date[2], date[1], date[0]].join('-')
        let time = new Date().toLocaleTimeString().split(" ")[0].split(":").map(x => {
            let n = x.toString();
            n = n.length < 2 ? `0${n}` : n;
            return n
        })
        time = [time[0], time[1], time[2]].join(":")
        this.date = dispatchState.date || date
        this.time = dispatchState.time || time
        this.riders = dispatchState.riders
        this.windSpeed = dispatchState.windSpeed === '' ? 0.0 : dispatchState.windSpeed
        this.windDegrees = dispatchState.windDegrees === '' ? 0 : dispatchState.windDegrees
        this.windsInstructor = dispatchState.windsInstructor
        this.btRadio = dispatchState.btRadio
        this.comment = dispatchState.comment
        this._id = dispatchState._id || null
    }
}

File: backend/models/dispatch.py
import pprint
from models._rider import Rider

class Dispatch:
    def __init__(self, wind_deg, wind_spd, bt, inst, date, time, comment="", riders={}, _id=None):
        self.riders = {'1': None, '2': None, '3': None, '4': None} | riders
        self.date = date
        self.time = time
        self._id = _id if _id else None
        self.wind_degrees = wind_deg
        self.wind_speed = wind_spd
        self.bt_radio = bt
        self.winds_instructor = inst
        self.comment = comment

    def __repr__(self):
        return pprint.pformat(self.as_dict())

    def as_dict(self):
        return {
            'riders': {line: rider.__dict__ for line, rider in self.riders.items()},
            'windDegrees': self.wind_degrees,
            'windSpeed': self.wind_speed,
            'date': self.date.isoformat(),
            'time': self.time.isoformat(),
            'btRadio': self.bt_radio,
            'windsInstructor': self.winds_instructor,
            'comment': self.comment,
            '_id': str(self._id)
        }

    @classmethod
    def from_dict(cls, d):
        """
        Dispatch factory from supplied dictionary obtained from database or frontend
        """
        return cls(
            wind_deg=d['windDegrees'],
            wind_spd=d['windSpeed'],
            bt=d['btRadio'],
            inst=d['windsInstructor'],
            date=d['date'],
            time=d['time'],
            comment=d['comment'],
            riders={line: Rider.from_dict(rider)
                    for line, rider in d['riders'].items()},
            _id=d['_id'] if '_id' in d.keys() else None
        )
